MaxText/train.py

```diff
@@ -480,14 +482,28 @@ def main(argv: Sequence[str]) -> None:
   config = pyconfig.config
   validate_train_config(config)
   os.environ["TFDS_DATA_DIR"] = config.dataset_path
+
+  if config.create_vertex_tensorboard and jax.process_index()==0:
+    tensorboard_url = vertex_tensorboard.setup(config)
+    if tensorboard_url is None:
+      raise ValueError("Unable to create Tensorboard and Experiment in Vertex AI.")
+    max_logging.log(f"View your Vertex AI Tensorboard at: {tensorboard_url}")
+  vertex_uploader_flag = False
+  if (os.environ.get("UPLOAD_DATA_TO_TENSORBOARD") or config.upload_data_to_tensorboard) and jax.process_index() == 0:
+    vertex_uploader_flag = vertex_tensorboard.upload_data(config)
+
   debug_config = debug_configuration.DebugConfig(
     stack_trace_config = stack_trace_configuration.StackTraceConfig(
       collect_stack_trace = config.collect_stack_trace,
       stack_trace_to_cloud = config.stack_trace_to_cloud,
       stack_trace_interval_seconds = config.stack_trace_interval_seconds))
   diagnostic_config = diagnostic_configuration.DiagnosticConfig(debug_config)
   with diagnostic.diagnose(diagnostic_config):
-    train_loop(config)
+    try:
+      train_loop(config)
+    finally:
+      if vertex_uploader_flag:
+        uploader.stop_upload_to_tensorboard()
```

Comment:
Shouldn't this be handle with a destructor on uploader?

Reply:
Do you mean using `__del__` in python? I believe it is not guaranteed that the object will be destroyed: https://docs.python.org/3/reference/datamodel.html#object.__del__. Therefore, I have explicitly added a method to destroy the thread. Also, this design was followed to align with the Vertex AI APIs: https://cloud.google.com/vertex-ai/docs/experiments/tensorboard-upload-existing-logs. 

What are your thoughts?

Reply:
I don't understand. isn't the Python3 doc saying that in a teardown event we might not call your destructor? But in that event do you want to teardown this? (Recall in general we can segfault so there is no promise.)

What is this `stop_upload_to_tensorboard` doing exactly?

Reply:
Added a new `VertexTensorboardManager` class that will take care of destroying the thread when the program ends.